Implement ConfigService for getting and setting module configs

hikinggrass opened the pull request.

hikinggrass committed:
Implement ConfigService for getting and setting module configs

Add userconfig storage backend

Add access section to config schema

throw exception when Unknown Datatype is encountered

Restructure get_module_config to use types from config service header

Use db for setting config values when in use, store config access information in db as well

Add Unknown entries to config service related enums

Add more meaningful exceptiption messages

Add get_config_value to ConfigService

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Ensure that module impl id is always set since it it NOT NULL in the db

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Add TODO stating that values written to the db that returned a
RebootRequired status will already be returned by a get_config_value()
call

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Add MODULE_IMPLEMENTATION_ID constant to config service header

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Update version to 0.22.0

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
clang-format

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Revert back to existing module_implementation_id handling

This fixes the failing tests

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Add missing mutability field to everestrs manifest handling

This fixes the everestrs tests

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

Pietfried commented on include/utils/config/types.hpp:
```diff
@@ -160,6 +177,21 @@ struct ConfigurationParameter {
+
+struct Access {
+    std::optional<ConfigAccess> config;
+};
```
Some brief doxygen would help, also for the members, it was not immediately clear to me which type is used for which use case

hikinggrass replied:
Added doxygen for the new types

Pietfried commented on include/utils/config_service.hpp:
```diff
@@ -0,0 +1,171 @@
+};
+
+struct SetResponse {
+    SetResponseStatus status = SetResponseStatus::Rejected;
```
this could be extended with an optional `reason` or `status_info`

hikinggrass replied:
I've added `status_info` where applicable

Pietfried commented on include/utils/config_service.hpp:
```diff
@@ -0,0 +1,171 @@
+    std::optional<everest::config::ConfigurationParameterIdentifier> identifier; ///< Used for GetType::Value
+};
+
+struct GetResponse {
```
this could be extended with an optional reason or status_info

Pietfried commented on include/utils/config_service.hpp:
```diff
@@ -0,0 +1,171 @@
+
+class ConfigServiceClient {
+public:
+    ConfigServiceClient(std::shared_ptr<MQTTAbstraction> mqtt_abstraction, const std::string& module_id,
+                        const std::unordered_map<std::string, std::string>& module_names);
```
also here some documentation would help, what are the `module_names` specifying here?

hikinggrass replied:
Added doxygen to explain the reason for the `module_names` here

Pietfried commented on include/utils/types.hpp:
```diff
@@ -135,6 +135,14 @@ namespace Everest {
+struct MQTTRequest {
+    std::string response_topic;
+    QOS qos = QOS::QOS2;
+    std::chrono::milliseconds timeout = std::chrono::milliseconds(5000); // FIXME: default?
```
yes, you can define a constant default

Pietfried commented on include/utils/config/types.hpp:
```diff
@@ -160,6 +177,21 @@ struct ConfigurationParameter {
+struct ConfigAccess {
+    bool allow_global_read = false;
+    bool allow_set_read_only = false;
+    std::map<std::string, everest::config::ModuleConfigAccess> modules;
```
I guess the key of the map is the `module_id` ? should be documented

Pietfried commented on lib/config/userconfig_storage.cpp:
```diff
@@ -0,0 +1,100 @@
+namespace everest::config {
+
+UserConfigStorage::UserConfigStorage(const fs::path& user_config_path) : user_config_path(user_config_path) {
+    // TODO: try-catch
```
should be resolved

hikinggrass replied:
Added more error handling

Pietfried commented on lib/config/userconfig_storage.cpp:
```diff
@@ -0,0 +1,100 @@
+    return GenericResponseStatus::Failed;
+}
+GenericResponseStatus UserConfigStorage::wipe() {
+    // FIXME: error handling
```
should be resolved

Pietfried commented on lib/config_service.cpp:
```diff
@@ -0,0 +1,588 @@
+    return get_response;
+}
+
+std::tuple<GetResponse, ResponseStatus> handle_get_config_value(const GetRequest& get_request,
```
Could the ResponseStatus be part of the GetResponse?

hikinggrass replied:
Changed to return type for setting and getting

hikinggrass committed:
Cache existing config value if set_config_value needs a reboot to complete

This prevents get config value calls from already returning a value that is not supposed to be valid yet

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Since there are already database migrations file in main starting with "2", prefix the new migrations with "3"

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

Pietfried commented on include/utils/config/types.hpp:
```diff
@@ -161,6 +178,21 @@ struct ConfigurationParameter {
 };
+
+struct ModuleConfigAccess {
+    bool allow_set_read_only = false;
```
nit:
```suggestion
    bool allow_set_read_only = false;
    bool allow_read = false;
```

hikinggrass replied:
I made this a bit more fine-grained, it can now look like this:

```yaml
    access:
      config:
        allow_global_read: true
        allow_global_write: false
        allow_set_read_only: false
        modules:
          auth:
            allow_read: true
            allow_write: true
            allow_set_read_only: true
```

hikinggrass committed:
Added missing doxygen comments

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Address some clang-tidy warnings

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Re-use a constant for the MQTTRequest timeout

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Add error handling

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Removed unused "remap" entry in config schema

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Modify return types to include status information

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Fix set response not being set in handle_set_request

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Make config access a bit more flexible by adding explicit allow read/write setting like this:

access:
  config:
    allow_global_read: true
    allow_global_write: false
    allow_set_read_only: false
    modules:
      some_module_id:
        allow_read: true
        allow_write: true
        allow_set_read_only: true

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

Pietfried commented on include/utils/config/types.hpp:
```diff
@@ -179,18 +179,29 @@ struct ConfigurationParameter {
-    bool allow_global_read = false;
-    bool allow_set_read_only = false;
-    std::map<std::string, everest::config::ModuleConfigAccess> modules;
+    bool allow_global_read = false;   ///< Allow this module to read the config items of other all other modules
```
```suggestion
    bool allow_global_read = false;   ///< Allow this module to read the config items of all other modules
```

Pietfried commented on include/utils/config/types.hpp:
```diff
@@ -179,18 +179,29 @@ struct ConfigurationParameter {
-    bool allow_set_read_only = false;
-    std::map<std::string, everest::config::ModuleConfigAccess> modules;
+    bool allow_global_read = false;   ///< Allow this module to read the config items of other all other modules
+    bool allow_global_write = false;  ///< Allow this module to write the config items of other all other modules
```
```suggestion
    bool allow_global_write = false;  ///< Allow this module to write the config items of all other modules
```

Pietfried commented on include/utils/config_service.hpp:
```diff
@@ -6,91 +6,132 @@
-    everest::config::SetConfigStatus
-    set_config_value(const everest::config::ConfigurationParameterIdentifier& identifier, const std::string& value);
+    /// \brief Sets the config \p value associated with the \p identifier
+    /// \returns a resukt containing status and potential error information
```
```suggestion
    /// \returns a result containing status and potential error information
```

Pietfried reviewed (approved):
lgtm

hikinggrass committed:
Fix spelling of comments

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>
